src/rust/engine/graph/README.md

```diff
@@ -0,0 +1,41 @@
+
+In the simplest case, if the `Node` is already `Completed`, the `Graph` immediately returns the Node's result value from the `Entry`. If the `Node` is currently `Running`, the caller will be added as a waiter on the `Entry`, and pushed a value when it is ready. Finally, if the `Node` is `NotStarted`, its `run` method is launched (concurrently on a `tokio` task) and the `Entry` moves to the `Running` state.
+
+Continuing the example above: the `Node` implementation for a `@rule` uses information (pre-computed by the `RuleGraph`) about the `@rule`'s signature to [request the values of `Node`s for the positional arguments of the `@rule`](https://github.com/pantsbuild/pants/blob/01372719f7e6f2a3aa0b9f3ce6909991388071ca/src/rust/engine/src/nodes.rs#L941-L966), and then runs the `@rule` function. If the result of running the `@rule` is a generator value, the `Node` then continues running to request more `Node` values for `Get`s: otherwise it completes with the result value.
+
+## Invalidation
+
+When a `Node` requests the value of another `Node`, it does so via `NodeContext::get` ([implemented for the `Context` struct](https://github.com/pantsbuild/pants/blob/01372719f7e6f2a3aa0b9f3ce6909991388071ca/src/rust/engine/src/context.rs#L418-L456) in production usage), which ends up calling `Graph::get`. Unlike an external caller, when a `Node` requests another `Node`, an edge is created in the `Graph` to record the dependency between the two `Node` instances. This tracking is used to implement invalidation of `Node`s.
+
+At any point (perhaps concurrently with `Node`s running), external threads may call [Graph::invalidate_from_roots](https://github.com/pantsbuild/pants/blob/01372719f7e6f2a3aa0b9f3ce6909991388071ca/src/rust/engine/graph/src/lib.rs#L401-L407) to indicate that the result values of particular `Node` instances are no longer valid for some reason. Because the identity of a `Node` is immutable (see the "Inputs and Identity" section above), invalidation almost always needs to occur because some information that the `Node` fetched using the `Context` (or direct use of syscalls, etc) is no longer valid. In production usage, invalidation occurs due to filesystem changes: a file watching thread [calls invalidate_from_roots](https://github.com/pantsbuild/pants/blob/01372719f7e6f2a3aa0b9f3ce6909991388071ca/src/rust/engine/src/context.rs#L351-L362) for `Node` types that are associated with syscalls to read files when it suspects that files on disk have changed.
+
+Confusingly, "root" in the context of `invalidate_from_roots` means a root in the dependee graph: calling the method for a root `Node` will clear the value of that `Node`, and mark the transitive dependees of that `Node` "dirty". The cleared `Node` will definitely re-run when it is next requested (it goes back to the `NotStarted` state), but dirty `Node`s will only re-run if they cannot be "cleaned": see below.
```

Comment:
Should clarify that we never delete nodes, so clearing a `Node` just means setting its `Entry` to `NotStarted`.